File: src/main/java/ir/alireza/sadeghi/shiraz/AI/GameTree.java
package ir.alireza.sadeghi.shiraz.AI;

import java.util.ArrayList;
import java.util.List;

public class GameTree {
	Tree<GameState> gameTree;

	public GameTree(Node<GameState> root) {
		gameTree = new Tree<>(root);
	}
	
	public void setNewRoot(Node<GameState> newRoot) {
		DeleteLayers.deleteEverythingAbove(newRoot, this.gameTree);
	} 
	
	public void buildFullTree(int layers) {
		//builds the entire tree we need (can be a LOT of nodes)
		gameTree.getRoot().children.clear();
		
		List<Node<GameState>> expanding = new ArrayList();
		expanding.add(gameTree.getRoot());

		//create layers
		for (int i = 1; i < layers + 1; i++) {
			AddNodes.addForMultiple(expanding);
			expanding = findAtDepth(i);
		}
	}
	
	//finds all the nodes at a certain depth
	public List<Node<GameState>> findAtDepth(int depth){
		Node<GameState> current = gameTree.getRoot();
		List<Node<GameState>> currentKids = new ArrayList();
		List<Node<GameState>> newKids = new ArrayList();
		
		currentKids.add(current);
		
		for (int i = 0; i < depth; i++) {
			for(int j = 0; j < currentKids.size(); j++) {
				if (currentKids.get(j).children !=null) {
					 newKids.addAll(currentKids.get(j).children);
				}
			}
			currentKids.clear();
			currentKids.addAll(newKids);
			newKids.clear();
		}
		return currentKids;
	}

	public Node<GameState> getRoot() {
		return gameTree.root;
	}
}

File: src/main/java/ir/alireza/sadeghi/shiraz/AI/GameTree.kt
package ir.alireza.sadeghi.shiraz.ai

class GameTree(root: Node<GameState?>?) {
    var gameTree: Tree<GameState?>

    init {
        gameTree = Tree(root)
    }

//    fun setNewRoot(newRoot: Node<GameState?>) {
//        DeleteLayers.deleteEverythingAbove(newRoot, gameTree)
//    }

    fun buildFullTree(layers: Int) {
        //builds the entire tree we need (can be a LOT of nodes)
        gameTree?.root?.children?.clear()
        var expanding: MutableList<Node<GameState?>?> = ArrayList()
        expanding.add(gameTree?.root)

        //create layers
        for (i in 1 until layers + 1) {
            AddNodes.addForMultiple(expanding)
            expanding = findAtDepth(i)
        }
    }

    //finds all the nodes at a certain depth
    private fun findAtDepth(depth: Int): MutableList<Node<GameState?>?> {
        val current = gameTree.root
        val currentKids: MutableList<Node<GameState?>?> = ArrayList()
        val newKids: MutableList<Node<GameState?>?> = ArrayList()
        currentKids.add(current)
        for (i in 0 until depth) {
            for (j in currentKids.indices) {
                if (currentKids[j]!!.children != null) {
                    newKids.addAll(currentKids[j]!!.children!!)
                }
            }
            currentKids.clear()
            currentKids.addAll(newKids)
            newKids.clear()
        }
        return currentKids
    }

    val root: Node<GameState?>?
        get() = gameTree.root
}